### PR TITLE
ls: handle broken symlink

### PR DESCRIPTION
**Pull request by pared**
Fixes: #7120

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/repo/ls.py [hunk 1]
```diff
@@ -66,16 +66,28 @@ def onerror(exc):
...
+
+        except FileNotFoundError:
+            # if walk found it but there is no metadata, its broken symlink
+            if not dvc_only:
+                ret[path] = {
+                    "isout": False,
+                    "isdir": False,
+                    "isexec": False,
+                }
+
```
Comment (efiop) [inline dvc/repo/ls.py:92]: Note that you are ignoring FileNotFoundError if `dvc_only`, which could be a dangerous bug. Also duplicating code, so how about smth like:

```
        try:
            metadata = fs.metadata(info)
        except FileNotFoundError:
            if dvc_only:
                raise
                
            metadata = Metadata()
            
        if metadata.output_exists or not dvc_only:
            path = (
                fs.path.name(fs_path)
                if fs_path == info
                else fs.path.relpath(info, fs_path)
            )
            ret[path] = {
                "isout": metadata.is_output,
                "isdir": metadata.isdir,
                "isexec": metadata.is_exec,
            }
```

**Comment by pared** [comment]
> Btw, how does CLI ls handle broken symlinks? Should we colorize it too? If so, how hard would that be?

CLI ls handles it the same way as symlinks - we get a color for symlink. 
I think that we should not consider that in that in this change: so far we haven't been differentiating between files/dirs and links. 

We could consider implementing it. The problem with that is that currently, we are differentiating the colors by determining whether the target is output/dir/executable. If none of the above, we use an extension to determine it. So far there has been no talking about links. In order to provide support, we would need to pass `islink` returned by `fsspec.info` up to the Metadata and leverage it to assign a color.

The question is whether we should do that. CLI ls is about inspecting the filesystem so it makes sense to mark the links. However the DVC is about our ML project - the most important is to differentiate between tracked and non tracked assets. I think w should not expand that. Even now, the behavior might be a bit counterintuitive. For example: if our pipeline will be producing executable, it will still be colored as output, because it is the first check we do.  I think in this case we should not follow `ls` style.

As an alternative, we could consider how to "mix" colors depending on the multiple properties of a given asset.

File: dvc/repo/ls.py [hunk 2]
```diff
@@ -66,7 +69,19 @@ def onerror(exc):
 
     ret = {}
     for info in infos:
-        metadata = fs.metadata(info)
+        path = (
+            fs.path.name(fs_path)
+            if fs_path == info
+            else fs.path.relpath(info, fs_path)
+        )
```
Comment (efiop) [inline dvc/repo/ls.py:76]: Doesn't look used until defined below
```suggestion
```

File: dvc/repo/ls.py [hunk 3]
```diff
@@ -66,7 +69,19 @@ def onerror(exc):
...
+        try:
+            metadata = fs.metadata(info)
+        except FileNotFoundError:
+            if dvc_only:
+                raise
+
+            metadata = Metadata(fs_path, repo)
```
Comment (efiop) [inline dvc/repo/ls.py:83]:
```suggestion
            metadata = Metadata(info, repo)
```

File: dvc/repo/ls.py [hunk 4]
```diff
@@ -66,7 +69,19 @@ def onerror(exc):
...
+            fs.path.name(fs_path)
+            if fs_path == info
+            else fs.path.relpath(info, fs_path)
+        )
+        try:
+            metadata = fs.metadata(info)
+        except FileNotFoundError:
+            if dvc_only:
+                raise
+
```
Comment (efiop) [inline dvc/repo/ls.py:82]:
```suggestion
            # broken symlink
```

File: dvc/repo/ls.py [hunk 5]
```diff
@@ -66,7 +69,15 @@ def onerror(exc):
 
     ret = {}
     for info in infos:
-        metadata = fs.metadata(info)
+        try:
+            metadata = fs.metadata(info)
+        except FileNotFoundError:
+            if dvc_only:
+                raise
```
Comment (pared) [inline dvc/repo/ls.py:76]: With the current approach (creating empty Metadata) I think it makes sense to remove this check.
it shouldn't matter if we use `dvc_only` or not. In my initial version lack of this was a bug, silently skipping the error, but now it seems ok to remove it. Users should not be punished for a specific check. If there is a broken symlink, other commands will make sure of erroring out

Comment (pared) [reply dvc/repo/ls.py:76]: Also, in the case of `dvc_only` files, we should never get here: repo_fs will be able to get the metadata even for the removed file

**Review by efiop — APPROVED** [review]

**Pull request merged**